onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -0,0 +1,706 @@
+          (it->truenode->id.node_id == it->id.node_id)) {
+        ORT_THROW("One falsenode is pointing either to itself, either to another tree.");
+      }
+    } else
+      it->truenode = NULL;
+
+    coor.node_id = static_cast<int>(nodes_falsenodeids[i]);
+    found = idi.find(coor);
+    if (found == idi.end()) {
+      ORT_THROW("Unable to find node ", coor.tree_id, "-", coor.node_id, " (falsenode).");
+    }
+    if (coor.node_id >= 0 && coor.node_id < n_nodes_) {
+      it->falsenode = found->second;
+      if ((it->falsenode->id.tree_id != it->id.tree_id) ||
+          (it->falsenode->id.node_id == it->id.node_id)) {
+        ORT_THROW("One falsenode is pointing either to itself, either to another tree.");
+      }
+    } else
+      it->falsenode = NULL;
+  }
+
+  int64_t previous = -1;
+  for (i = 0; i < static_cast<size_t>(n_nodes_); ++i) {
+    if ((previous == -1) || (previous != nodes_[i].id.tree_id))
+      roots_.push_back(nodes_ + i);
+    previous = nodes_[i].id.tree_id;
+  }
+
+  TreeNodeElementId ind;
+  SparseValue<OTYPE> w;
+  for (i = 0; i < target_class_nodeids.size(); i++) {
+    ind.tree_id = static_cast<int>(target_class_treeids[i]);
+    ind.node_id = static_cast<int>(target_class_nodeids[i]);
+    if (idi.find(ind) == idi.end()) {
+      ORT_THROW("Unable to find node ", coor.tree_id, "-", coor.node_id, " (weights).");
+    }
+    w.i = target_class_ids[i];
+    w.value = target_class_weights[i];
+    idi[ind]->weights.push_back(w);
+  }
+
+  n_trees_ = roots_.size();
+  has_missing_tracks_ = false;
+  for (auto itm = nodes_missing_value_tracks_true.begin();
+       itm != nodes_missing_value_tracks_true.end(); ++itm) {
+    if (*itm) {
+      has_missing_tracks_ = true;
+      break;
+    }
+  }
+}
+
+template <typename ITYPE, typename OTYPE>
+void TreeEnsembleCommon<ITYPE, OTYPE>::compute(const Tensor* X, Tensor* Z, Tensor* label) const {
+  switch (aggregate_function_) {
+    case AGGREGATE_FUNCTION::AVERAGE:
+      compute_agg(
+          X, Z, label,
+          _AggregatorAverage<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+    case AGGREGATE_FUNCTION::SUM:
+      compute_agg(
+          X, Z, label,
+          _AggregatorSum<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+    case AGGREGATE_FUNCTION::MIN:
+      compute_agg(
+          X, Z, label,
+          _AggregatorMin<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+    case AGGREGATE_FUNCTION::MAX:
+      compute_agg(
+          X, Z, label,
+          _AggregatorMax<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+  }
+  ORT_THROW("Unknown aggregation function in TreeEnsemble.");
+}
```

Comment:
nit: putting this as the 'default' case in the switch statement would be more usual. could also include the value of aggregate_function_ in the error message.

Reply:
fixed